relay-ctrl-check.8
.TH relay-ctrl-check 8
.SH NAME
relay-ctrl-check \- check if the (SMTP) client is allowed to relay
.SH SYNOPSIS
.B relay-ctrl-check
.B program
[
.I arguments ...
]
.SH DESCRIPTION
Before executing
.BR program ,
.B relay-ctrl-check
checks if
.I $RELAYCLIENT
is set.  If not, it looks up the IP specified by
.I $TCPREMOTEIP
in the directory specified by
.IR $RELAY_CTRL_DIR .
If the IP file exists, and is not more than
.I $RELAY_CTRL_EXPIRY
seconds old (which defaults to 900 or 15 minutes), then
.I $RELAYCLIENT
is set to the value of
.I $RELAY_CTRL_RELAYCLIENT
(which defaults to an empty string).
The IP file is also parsed for environment variable settings which are exported to
.BR program .
Otherwise the IP file is removed and ignored.
If
.I $RELAY_CTRL_LOG_IPS
is set, this program will output a message reporting if it set
.IR $RELAYCLIENT ,
and why or why not.
If
.I $RELAY_CTRL_LOG_ENV
is set, all environment variable settings are output as well.
.SH "SEE ALSO"
relay-ctrl-age(8),
relay-ctrl-allow(8),
qmail-smtpd(8),
tcpserver(1)
.SH CAVEATS
.B relay-ctrl-check
loads environment variables from the IP file using a large but fixed
size buffer. Absurdly long variable settings will be silently ignored.


NEWS
-------------------------------------------------------------------------------
Changes in version 3.2

- Updated for bglibs v2.

- Fixed the authenticated test when used with Courier IMAP.

- Tweaked relay-ctrl-allow to only try to save a handle to the current
  working directory if it's going to execute another command later.

- Added support for logging environment settings in relay-ctrl-check.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 3.1.1

- Fixed the obvious bugs in relay-ctrl-udp (it didn't send an
  acknowledgement packet, among other things).

- Modified relay-ctrl-chdir to dup the opened directory file descriptor
  to the highest possible slot, to work around programs that close
  opened file descriptors (ie various Courier programs).

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 3.1

- relay-ctrl-allow now communicates the $USER and $DOMAIN variables to
  relay-ctrl-check.  relay-ctrl-send/relay-ctrl-udp don't handle this
  yet.

- Modified the function that creates IP files to first create a new
  temporary file and rename it to ensure the last-modified timestamp is
  updated.

- Added a relay-ctrl-chdir program, to open a directory so the
  relay-ctrl-* programs can inherit the opened FD to fchdir.  This
  allows all the programs to be executed non-setuid.

- Various portability bugs fixed in the libraries.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 3.0

- Removed all CDB management code in favour of doing the check in a new
  program, relay-ctrl-check, which is a front-end to qmail-smtpd.

- Added a client-server pair for transmitting relay authorization
  information over UDP.

- All configuration information is now read from environment variables.

- Running relay-ctrl-age is no longer *required*, but still recommended.

- The RPM no longer tries to add itself to any qmail control files.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 2.5

- Replaced most of the defines with run-time config files.  See the man
  page for relay-ctrl-age for more details.

- Removed check for standard input and standard output being sockets to
  allow relay-ctrl-allow to be used with SSL pipes.

- Check for the "::ffff:" IPv6 address prefix before validating address.

- Rules defined in the rules file now override the generated rules,
  instead of vice versa.
-------------------------------------------------------------------------------
Changes in version 2.0

- Now supports Courier IMAP as well as qmail-popup/qmail-pop3d.
-------------------------------------------------------------------------------


relay-ctrl-check.c
#include <bglibs/sysdeps.h>
#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <bglibs/msg.h>
#include <bglibs/systime.h>
#include "relay-ctrl.h"

const char program[] = "relay-ctrl-check";
const int msg_show_pid = 1;
static int log_ips = 0;
static int log_env = 0;

static const char* rc;
static long expiry;

static void load_env(int fd)
{
  static char buf[8192];
  int rd;
  char* ptr;
  char* end;
  int len;
  
  if ((rd = read(fd, buf, sizeof buf)) == -1)
    warn1sys("Could not read from opened IP file?!?");
  else if (rd > 0) {
    for (ptr = buf; rd > 0; ptr += len, rd -= len) {
      if ((end = memchr(ptr, 0, rd)) == 0) break;
      len = end - ptr + 1;
      if (memchr(ptr, '=', len) != 0) {
	if (log_env)
	  msg2("Setting ", ptr);
	if (putenv(ptr) == -1)
	  warn1("Could not set environment string");
      }
    }
  }
}

static void stat_ip(const char* ip)
{
  time_t now;
  struct stat s;
  int fd;
  if ((fd = open(ip, O_RDONLY)) == -1) {
    if (errno == ENOENT) {
      if (log_ips)
	msg3("IP ", ip, " not found");
    }
    else
      warn3sys("Could not open IP file '", ip, "'");
  }
  else {
    if (fstat(fd, &s) == -1)
      warn1sys("Could not fstat opened IP file?!?");
    else {
      now = time(0);
      if (s.st_mtime + expiry >= now) {
	if (log_ips)
	  msg3("IP ", ip, " found, setting $RELAYCLIENT");
	setenv("RELAYCLIENT", rc, 1);
	load_env(fd);
	close(fd);
      }
      else {
	if (log_ips)
	  msg3("Expired IP ", ip, " found, removing");
	unlink(ip);
      }
    }
  }
}

int main(int argc, char* argv[])
{
  const char* ip;
  const char* tmp;
  
  if (argc < 2) die1(1, "usage: relay-ctrl-check program [arguments]\n");
  log_ips = getenv("RELAY_CTRL_LOG_IPS") != 0;
  log_env = getenv("RELAY_CTRL_LOG_ENV") != 0;
  if (getenv("RELAYCLIENT") == 0) {
    expiry = 0;
    if ((tmp = getenv("RELAY_CTRL_EXPIRY")) != 0) expiry = atol(tmp);
    if (expiry <= 0) expiry = 900;
    if ((rc = getenv("RELAY_CTRL_RELAYCLIENT")) == 0) rc = "";
    if ((ip = getenv("TCPREMOTEIP")) == 0)
      warn1("$TCPREMOTEIP not set, not checking IP");
    else if (do_chdir(0))
      stat_ip(ip);
  }
  else {
    if (log_ips)
      msg1("$RELAYCLIENT already set, not checking IP");
  }
  execvp(argv[1], argv+1);
  die1(111, "execution of program failed!\n");
  return 111;
  argc = 0;
}
